### PR TITLE
support to delete undolog in back task

### PR DESCRIPTION
**Pull request by github-ygy**
### Ⅰ. Describe what this PR did
support to delete undolog in back task

### Ⅱ. Does this pull request fix one issue?

a feature，some case the undolog will not delete ,the pr can avoid the undo_log table so big

### CODE REVIEW

File: core/src/main/java/io/seata/core/protocol/transaction/UndoLogDeleteRequest.java [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+        byteBuffer.flip();
+        byte[] content = new byte[byteBuffer.limit()];
+        byteBuffer.get(content);
+        return content;
+    }
+
+    @Override
+    public boolean decode(ByteBuf in) {
+
+        if (in.readableBytes() < 1) {
```
Comment (xingfudeshi) [inline core/src/main/java/io/seata/core/protocol/transaction/UndoLogDeleteRequest.java]: How about writing this line like this:
if(in.readableBytes() < 2){
  return false;
}
this.branchType = BranchType.get(in.readByte());

File: core/src/main/java/io/seata/core/rpc/ChannelManager.java [hunk 2]
```diff
@@ -490,4 +490,24 @@ private static Channel tryOtherApp(ConcurrentMap<String, ConcurrentMap<String, C
...
+    /**
+     * get rm channels
+     *
+     * @return
+     */
+    public static Map<String,Channel> getRmChannels() {
+        if (RM_CHANNELS.isEmpty()) {
+            return null;
+        }
+        Map<String,Channel> channels = new HashMap(RM_CHANNELS.size());
```
Comment (xingfudeshi) [inline core/src/main/java/io/seata/core/rpc/ChannelManager.java]: ConcurrentMap

Comment (github-ygy) [reply core/src/main/java/io/seata/core/rpc/ChannelManager.java]: why?it thread safe.

File: core/src/main/java/io/seata/core/protocol/transaction/UndoLogDeleteRequest.java [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+    }
+
+    public void setBranchType(BranchType branchType) {
+        this.branchType = branchType;
+    }
+
+    @Override
+    public AbstractTransactionResponse handle(RpcContext rpcContext) {
+        handler.handle(this);
+        return null;
```
[the remark this answers is not in the document]
Comment (github-ygy) [reply core/src/main/java/io/seata/core/protocol/transaction/UndoLogDeleteRequest.java]: > why return null ?

async，wo don't need reply

File: core/src/main/java/io/seata/core/rpc/ChannelManager.java [hunk 4]
```diff
@@ -490,4 +490,24 @@ private static Channel tryOtherApp(ConcurrentMap<String, ConcurrentMap<String, C
...
+        for (String resourceId : RM_CHANNELS.keySet()) {
+            Channel channel = tryOtherApp(RM_CHANNELS.get(resourceId), null);
```
Comment (slievrly) [inline core/src/main/java/io/seata/core/rpc/ChannelManager.java]: why tryOtherApp？ tryOtherApp just get from RM_CHANNELS, loop is also getting from  RM_CHANNELS  and your map can only save one channel for a resource

File: server/src/main/java/io/seata/server/coordinator/DefaultCoordinator.java [hunk 5]
```diff
@@ -104,6 +108,11 @@
      */
     protected int timeoutRetryDelay = CONFIG.getInt(ConfigurationKeys.TIMEOUT_RETRY_DELAY, 5);
 
+    /**
+     * The Timeout retry delay.
+     */
+    protected int transactionUndologDeleteDelay = CONFIG.getInt(ConfigurationKeys.TRANSACTION_UNDO_LOG_DELETE_DELAY, 24 * 60);
```
Comment (slievrly) [inline server/src/main/java/io/seata/server/coordinator/DefaultCoordinator.java]: best to indicate the time unit

Comment (github-ygy) [reply server/src/main/java/io/seata/server/coordinator/DefaultCoordinator.java]: same with committingRetryDelay,asynCommittingRetryDelay,timeoutRetryDelay, default is SECONDS

File: core/src/main/java/io/seata/core/protocol/transaction/UndoLogDeleteRequest.java [hunk 6]
```diff
@@ -0,0 +1,139 @@
...
+            byte[] bs = resourceId.getBytes(UTF8);
+            byteBuffer.putShort((short)bs.length);
+            if (bs.length > 0) {
+                byteBuffer.put(bs);
+            }
+        } else {
+            byteBuffer.putShort((short)0);
+        }
+        //3.save days
+        byteBuffer.putInt(saveDays);
```
[the remark this answers is not in the document]
Comment (github-ygy) [reply core/src/main/java/io/seata/core/protocol/transaction/UndoLogDeleteRequest.java]: done

File: core/src/main/java/io/seata/core/rpc/ChannelManager.java (hunk 4 above)
Comment (github-ygy) [reply core/src/main/java/io/seata/core/rpc/ChannelManager.java]: the method i have modified，when the param applitioncation is null，will find one avtive channel from all application.

**Comment by github-ygy** [comment]
> Please add the configuration item to the configuration file.

done

**Pull request merged**